### ISSUE TITLE
`useless_concat` false positive

### ISSUE THREAD

**Issue by gabi-250** [rust-lang/rust-clippy#15223]
### Summary

The `useless_concat` lint seems to incorrectly trigger when `concat!` is called within a `macro_rules` block generated by a particular [derive macro](https://docs.rs/derive-deftly/1.1.0/derive_deftly/doc_implementation/index.html#introduction).  

### Lint Name

useless_concat

### Reproducer

I haven't had a chance to come up with a repro that doesn't involve an external dependency on [derive-deftly](https://crates.io/crates/derive-deftly) (but I'll try to add one later).   

I tried this code:

```rust
// cargo add derive-deftly
use derive_deftly::{define_derive_deftly, Deftly};

define_derive_deftly! {
    ListVariants:

    impl $ttype {
        fn list_variants() {
            // Display all the variant names
            $(
                println!("{}", concat!("variant: ", stringify!($vname)));
            )
        }
    }
}

#[derive(Deftly)]
#[derive_deftly(ListVariants)]
#[allow(dead_code)]
enum Enum {
    Unit,
    Struct { a: u8, b: u16 },
    Tuple(u8, u16),
}

fn main() {
    Enum::list_variants();
}
```

I saw this happen:

```
warning: unneeded use of `concat!` macro
  --> src/main.rs:10:32
   |
10 |                 println!("{}", concat!("variant: ", stringify!($vname)));
   |                                ^^^^^^^^ help: replace with: `"variant: Unit"`
...
16 | #[derive(Deftly)]
   |          ------ in this derive macro expansion
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_concat
   = note: `#[warn(clippy::useless_concat)]` on by default
   = note: this warning originates in the macro `derive_deftly_template_ListVariants` which comes from the expansion of the derive macro `Deftly` (in Nightly builds, run with -Z macro-backtrace for more info)

warning: unneeded use of `concat!` macro
  --> src/main.rs:10:32
   |
10 |                 println!("{}", concat!("variant: ", stringify!($vname)));
   |                                ^^^^^^^^ help: replace with: `"variant: Struct"`
...
16 | #[derive(Deftly)]
   |          ------ in this derive macro expansion
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_concat
   = note: this warning originates in the macro `derive_deftly_template_ListVariants` which comes from the expansion of the derive macro `Deftly` (in Nightly builds, run with -Z macro-backtrace for more info)

warning: unneeded use of `concat!` macro
  --> src/main.rs:10:32
   |
10 |                 println!("{}", concat!("variant: ", stringify!($vname)));
   |                                ^^^^^^^^ help: replace with: `"variant: Tuple"`
...
16 | #[derive(Deftly)]
   |          ------ in this derive macro expansion
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_concat
   = note: this warning originates in the macro `derive_deftly_template_ListVariants` which comes from the expansion of the derive macro `Deftly` (in Nightly builds, run with -Z macro-backtrace for more info)

warning: `useless-concat-fp` (bin "useless-concat-fp") generated 3 warnings (run `cargo clippy --fix --bin "useless-concat-fp"` to apply 3 suggestions)
```

I expected to see this happen:
no emitted warnings

### Version

```text
rustc 1.90.0-nightly (b03b3a7ec 2025-06-26)
binary: rustc
commit-hash: b03b3a7ec92682be2917540b679478d41c95a30c
commit-date: 2025-06-26
host: x86_64-unknown-linux-gnu
release: 1.90.0-nightly
LLVM version: 20.1.7
```
